File: week-17/pushedrumex/길 찾기 게임.py
from collections import defaultdict

def solution(nodeinfo):
    N = len(nodeinfo)
    y_dic = defaultdict(list)
    node_dic = {}
    for i, node in enumerate(nodeinfo):
        x, y = node
        y_dic[y].append(x)
        node_dic[(x, y)] = i+1
    
    tree = [[0, 0, 0] for _ in range(100_000 + 1)]
    ys = sorted(y_dic.keys())
    yN = len(ys)

    for i in range(yN-1, 0, -1):
        print(y_dic[ys[i]])
        print(y_dic[ys[i-1]])
        for parent in y_dic[ys[i]]:
            for child in y_dic[ys[i-1]]:
                if parent < tree[parent][0]:
                    if child < parent:
                        tree[parent][1] = child
                        tree[child][0] = parent
                    elif child < tree[parent][0]:
                        tree[parent][2] = child
                        tree[child][0] = parent
                else:
                    if child > parent:
                        tree[parent][2] = child
                        tree[child][0] = parent
                    elif child > tree[parent][0]:
                        tree[parent][1] = child
                        tree[child][0] = parent
    answer = [[]]
    return answer

File: week-17/dahyen0o/길-찾기-게임.java
import java.util.*;

class Solution {

    static class Node {
        final int num;
        final int[] coord;
        Node left, right;

        Node(int num, int[] coord) {
            this.num = num;
            this.coord = coord;
            this.left = null;
            this.right = null;
        }
    }

    private static final Map<String, Integer> toNum = new HashMap<>();

    private static int h = 0; // travel용 history index

    public int[][] solution(final int[][] nodeinfo) {
        // toNum 저장
        for(int num = 0; num < nodeinfo.length; num++) {
            toNum.put(toHash(nodeinfo[num]), num + 1);
        }

        // y축 기준으로 nodeinfo 정렬
        Arrays.sort(nodeinfo, (o1, o2) -> {
            final int y = Integer.compare(o2[1], o1[1]);
            if(y != 0) return y;
            return Integer.compare(o1[0], o2[0]);
        });

        // 트리 생성
        final Node root = new Node(toNum.get(toHash(nodeinfo[0])), nodeinfo[0]);

        for(int i = 1; i < nodeinfo.length; i++) {
            final int[] curr = nodeinfo[i];
            final int currNum = toNum.get(toHash(curr));

            final Node parent = findParent(root, curr);
            if(parent.coord[0] > curr[0]) { // left child
                parent.left = new Node(currNum, curr);
            } else { // right child
                parent.right = new Node(currNum, curr);
            }
        }

        // 트리 순회
        final int[][] answer = new int[2][nodeinfo.length];
        // 전위 순회
        h = 0;
        preTravel(root, answer[0]);
        // 후위 순회
        h = 0;
        postTravel(root, answer[1]);

        return answer;
    }

    private String toHash(final int[] coord) {
        return String.format("%d %d", coord[0], coord[1]);
    }

    private Node findParent(final Node curr, final int[] coord) {
        if(coord[0] < curr.coord[0]) { // left child
            if(curr.left != null) return findParent(curr.left, coord);
            return curr;
        }
        else { // right child
            if(curr.right != null) return findParent(curr.right, coord);
            return curr;
        }
    }

    private void preTravel(final Node curr, final int[] history) {
        if(curr == null) return;

        history[h++] = curr.num;

        preTravel(curr.left, history);
        preTravel(curr.right, history);
    }

    private void postTravel(final Node curr, final int[] history) {
        if(curr == null) return;

        postTravel(curr.left, history);
        postTravel(curr.right, history);
        history[h++] = curr.num;
    }
}
